core/src/main/scala/akka/kafka/internal/ConsumerProgressTracking.scala

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <https://softwaremill.com>
+ * Copyright (C) 2016 - 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.kafka.internal
+import akka.annotation.InternalApi
+import org.apache.kafka.clients.consumer.{Consumer, ConsumerRecords, OffsetAndMetadata}
+import org.apache.kafka.common.TopicPartition
+
+import scala.jdk.CollectionConverters._
+
+/**
+ * Maintain our own OffsetAndTimestamp which can tolerate negative timestamps, which happen for old clients that
+ * don't set timestamp explicitly.
+ */
+final case class SafeOffsetAndTimestamp(offset: Long, timestamp: Long)
+
+/**
+ * Listen for changes to the consumer assignments.
+ */
+@InternalApi
+trait ConsumerAssignmentTrackingListener {
+  def revoke(revokedTps: Set[TopicPartition]): Unit = {}
+  def assignedPositions(assignedTps: Set[TopicPartition], assignedOffsets: Map[TopicPartition, Long]): Unit = {}
+}
+
+/**
+ * Track the current state of the consumer: what offsets it has requested, received and committed, filtering by the
+ * current assignments to the consumer. When a partition is assigned to the consumer for the first time, its
+ * assigned offset is the current position of the partition (uses underlying Kafka Consumer to leverage the
+ * configured offset-reset policy).
+ */
+@InternalApi
+trait ConsumerProgressTracking extends ConsumerAssignmentTrackingListener {
+  def commitRequested: Map[TopicPartition, OffsetAndMetadata] = null
+  def receivedMessages: Map[TopicPartition, SafeOffsetAndTimestamp] = null
+  def committedOffsets: Map[TopicPartition, OffsetAndMetadata] = null
+
+  def commitRequested(offsets: Map[TopicPartition, OffsetAndMetadata]): Unit = {}
+  def received[K, V](records: ConsumerRecords[K, V]): Unit = {}
+  def committed(offsets: java.util.Map[TopicPartition, OffsetAndMetadata]): Unit = {}
+  def assignedPositionsAndSeek(assignedTps: Set[TopicPartition],
+                               consumer: Consumer[_, _],
+                               positionTimeout: java.time.Duration): Unit = {}
+  def addProgressTrackingCallback(callback: ConsumerAssignmentTrackingListener): Unit = {}
+}
+
+@InternalApi
+object ConsumerProgressTrackerNoop extends ConsumerProgressTracking {}
+
+/**
+ * Track the progress/state of the consumer. We generally try to be 'fast' with handling the partitions we track;
+ * most of the smarts are expected to be handled outside this class. For example, we will update any offsets to
+ * commit - [[commitRequested]] - without regard for what had previously been assigned
+ * or revoked from the consumer. Thus, care should be taken when managing state of the consumer and making updates.
+ *
+ * The only case we try and be "smart" is during [[received]], where we will filter out offsets that have not been
```

Comment:
@seglo when going through the behavior in tests, this stood out as a bit strange to me. Either feels like we should be smart for all entry to this class or not smart at all (none of this filter-sometimes behavior). I prefer to be 'dumb' here if we can and push the logic into the KafkaConsumerActor if we can. If so, then we could do the "is-assigned" filtering where we also do the reset protection (https://github.com/akka/alpakka-kafka/pull/1299/files#diff-c05b458cf33e1bf97dadc6ea4b01c56daf01a6ddb72bd207c62a7d8331441d60R617 - line 617), and drop the SourceLogicBuffer. 

However, it is quite hard to _not_ do the filtering either here or in the actor as we want to we might receive offsets that are not assigned anymore due to the nature of poll results. In that case, we would want to just filter out everything that is not assigned for all the methods here.

WDYT?

Reply:
> @seglo when going through the behavior in tests, this stood out as a bit strange to me. Either feels like we should be smart for all entry to this class or not smart at all (none of this filter-sometimes behavior). I prefer to be 'dumb' here if we can and push the logic into the KafkaConsumerActor if we can.

Good point. I'm leaning towards the "dumb" behaviour as well. What would you move to `KafkaConsumerActor` if we went this route?

> If so, then we could do the "is-assigned" filtering where we also do the reset protection (#1299 (files) - line 617), and drop the SourceLogicBuffer.

The `SourceLogicBuffer` filtering is important because a buffer exists between the `KafkaConsumerActor` and the first user stage. If partitions are revoked for messages already in the buffer then we'll send late messages, so it would be preferred to keep that logic as well. I'm not sure I understood your point correctly though.

Reply:
Apologies, was on vacation last week....

I also seem to have forgotten the cleverness i was considering when i wrote that comment. However, im realizing that there might be a subtle bug here - if we haven't committed the partition yet, we won't have it in `requestedOffsetsImpl` (which tracks the commits that have been requested...and clearly needs a slight rename) which means we won't track it yet. That is a little unexpected behavior from the outside - you would think received would just track things we receive. 

All of this to say, we should probably _not_ try and be clever here (clearly its hard) and just eat the miniscule amount of overhead of tracking some more partitions; its a topic name, partition ID and a long value for each, not really that bad in terms of memory when compared with all the other moving parts here. If its does become bad (too much memory) we could have a chaining map that does topic -> partition -> offset...but that also seems overkill at the moment.

Reply:
And now that i look more closely, if we just track the currently assigned partitions, we can filter on that instead. Since revokes and assignments are synchronous, we should be able to use that as our filtering mechanism; which to me makes at least logical sense from the outside.

Reply:
Looks ok to me.